pkg/updater/download.go

```diff
@@ -116,41 +193,142 @@ func extractTarGz(archivePath string, destinationPath string) error {
 		if err != nil {
 			return fmt.Errorf("could not read tar header: %w", err)
 		}
+
+		// Skip the root directory
+		if header.Name == "./" {
+			continue
+		}
+
 		target := filepath.Join(destinationPath, header.Name)
+
+		// Check for zip-slip attacks
+		if !strings.HasPrefix(target, filepath.Clean(destinationPath)+string(os.PathSeparator)) {
+			return fmt.Errorf("tar entry %s is trying to escape the destination directory", header.Name)
+		}
+
+		// Extract element depending on its type
 		switch header.Typeflag {
 		case tar.TypeDir:
 			err = os.MkdirAll(target, 0755)
 			if err != nil {
 				return fmt.Errorf("could not create directory: %w", err)
 			}
 		case tar.TypeReg:
-			err = extractTarGzFile(target, tr)
+			fileCount++
+			if fileCount > maxArchiveFileCount {
+				return errors.New("archive contains too many files")
+			}
+			err = extractTarFile(target, tr)
 			if err != nil {
 				return err // already wrapped
 			}
+		case tar.TypeLink, tar.TypeSymlink:
+			tarLinks = append(tarLinks, header)
 		default:
 			log.Warnf("Unsupported tar entry type %d for %s", header.Typeflag, header.Name)
 		}
 	}
+
+	// Process tar links afterwards as they may depend on other files being written
+	err = processTarLinks(0, destinationPath, tarLinks)
+	if err != nil {
+		return err // already wrapped
+	}
+
 	log.Debugf("Successfully extracted archive %s to %s", archivePath, destinationPath)
 	return nil
 }
 
-// extractTarGzFile extracts a file from a tar.gz archive.
+// extractTarFile extracts a file from a tar archive.
 // It is separated from extractTarGz to ensure `defer f.Close()` is called right after the file is written.
-func extractTarGzFile(targetPath string, reader io.Reader) error {
-	err := os.MkdirAll(filepath.Dir(targetPath), 0755)
+func extractTarFile(targetPath string, reader io.Reader) error {
+	err := os.MkdirAll(filepath.Dir(targetPath), 0o755)
 	if err != nil {
 		return fmt.Errorf("could not create directory: %w", err)
 	}
-	f, err := os.Create(targetPath)
+	f, err := os.Create(filepath.FromSlash(targetPath))
 	if err != nil {
 		return fmt.Errorf("could not create file: %w", err)
 	}
 	defer f.Close()
-	_, err = io.Copy(f, reader)
+
+	n, err := io.Copy(f, io.LimitReader(reader, maxArchiveFileSize))
 	if err != nil {
+		if errors.Is(err, io.EOF) && n == maxArchiveFileSize {
+			defer func() {
+				if err := os.Remove(targetPath); err != nil {
+					log.Errorf("Could not remove truncated file %q: %v", targetPath, err)
+				} else {
+					log.Debug("Removing truncated file %q", targetPath)
+				}
+			}()
```

Comment:
Nitpick 🥜: why is the above code wrapped in a defer block ?

Reply:
So that we can attempt to remove it after returning an error to the caller